### PR TITLE
Phase A: Reachy-side voice, embodiment, orchestration + E2E docs (M8–M13)

### PR DESCRIPTION
**Pull request by MacAttak**
## Summary

Reachy-side of Phase A — the app package, lifecycle docs, and networking guide. 18 commits, 27 files, ~2301 LOC. Every commit passed per-task spec + code-quality review; final cross-task sweep review surfaced two real resource-lifecycle bugs (I1, I2) that are also fixed here.

**M8 — Voice layer** (4 commits): `VoiceInterface` ABC + `MockVoice`, `OpenAIRealtimeVoice` (Realtime API), `WakeDetector` ABC + `MockWakeDetector`, `MuteGate`.

**M9 — Embodiment** (2): `MotionDriver` ABC + `ReachyMotionDriver`, `EmbodimentStateMachine` (IDLE / LISTENING / THINKING / MUTED transitions), face-tracking gaze helper.

**M10 — App orchestration** (2): async `DaemonClient` + `run_one_turn` single-turn conversation orchestrator.

**M11 — Entry point** (2): `ReachyDuckyApp` class + `main()`, HF Space metadata.

**M12 — E2E smoke doc** (3): `docs/testing/2026-04-21-phase-a-e2e-procedure.md` — full procedure with known gaps called out honestly.

**M13 — README** (3): Run, Development, Networking & auth (Tailscale-primary + bearer-token fallback).

**Resource-lifecycle fixes from sweep review** (2):
- **I1** — `VoiceInterface.start_turn()` is now an async context manager; Realtime websocket closes cleanly at turn end (was leaking per turn).
- **I2** — `DaemonClient` pools one `httpx.AsyncClient` across calls with `aclose()`; `ReachyDuckyApp._run_async` closes it in try/finally on shutdown (was N handshakes for N turns).

## Test plan

- [x] `uv run pytest -q` — 473 passed, 3 skipped (up from 464 pre-fix; +9 new tests for I1/I2 lifecycle)
- [x] `uv run mypy --strict daemon/src app/src menubar/src protocol/src` — clean (38 files)
- [x] `uv run mypy --strict daemon/tests protocol/tests menubar/tests app/tests` — clean (tests held to the same bar)
- [x] `uv run ruff check .` + `uv run ruff format --check .` — clean
- [x] `uv run bandit -ll -r daemon/src app/src menubar/src protocol/src` — clean
- [x] Lefthook pre-commit + pre-push hooks green on every commit
- [ ] Full E2E smoke procedure (`docs/testing/2026-04-21-phase-a-e2e-procedure.md`) run against real Reachy Mini hardware — **deferred until Known gaps §1 (wake word) + §3 (HF Space publish) close**

## Known gaps (tracked as follow-up issues)

Documented in `docs/testing/2026-04-21-phase-a-e2e-procedure.md` and the root README:

- Wake-word detection is a `MockWakeDetector` stub in Phase A; saying "Hey Ducky" does nothing today. A real detector is Phase B work.
- HF Space publish + on-robot install path is unverified; the YAML + README frontmatter are in place but nobody has walked the publish flow on hardware.
- Menu-bar mute is local UI only — does not propagate to the robot. When the audio pump lands, it should bind `MuteGate.set_muted(True)` at the same seam that fires `EmbodimentStateMachine.transition(MUTED)`.
- Menu-bar `_DAEMON_URL_DEFAULT` is hardcoded; not yet env-overridable.
- Cross-subpackage integration test (app ↔ daemon contract) doesn't exist; daemon schema drift won't be caught by CI until someone runs the E2E procedure by hand.

## Follow-ups from sweep review (to be opened as issues)

- **I3** — Repo-root `tests/` for cross-subpackage contract tests
- **I4** — Wake loop busy-spins at ~20Hz; should become `asyncio.Event` wait when real wake lands
- **I5** — HF Space metadata in two files (`reachy_mini_app.yaml` + README frontmatter); drift risk
- **I6** — `ReachyMotionDriver` SDK method names silenced with 6 `# type: ignore`; needs hardware-tier test
- **I7** — `_wake_triggered` deletes its `wake` param; wire or remove
- **M1** — `BrainRequest.include_tools` is wire-present but unused on both ends
- **M2** — No project-slug selector in the on-robot app
- **M3** — Mute coordination: bind `MuteGate.set_muted` to state-machine MUTED
- **M4** — `pyright` / `mypy` divergence on openai SDK private imports

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: app/src/reachy_ducky_app/conversation.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+        sm.transition(State.THINKING)
+        reply = await daemon.brain_query(user_text, project_slug=project_slug)
+
+        # Speak while in LISTENING posture — the robot looks at the user
+        # as Ducky replies, not while thinking.
+        sm.transition(State.LISTENING)
+        await turn.speak_text(reply.text)
+
+    sm.transition(State.IDLE)
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply app/src/reachy_ducky_app/conversation.py]: Fixed in 835b35b. Wrapped the orchestration body in `try/finally: sm.transition(State.IDLE)`, so state resets regardless of which await raises. Exception still propagates to the caller. Three new tests cover raise-from-`get_user_text`, raise-from-`brain_query`, and raise-from-`speak_text` — each asserts state is IDLE after the exception and that the exception reaches the caller via `pytest.raises`.

File: app/src/reachy_ducky_app/embodiment/gaze.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    ``mediapipe`` is imported lazily so this module stays importable on dev
+    machines where the runtime detector may not yet be exercised.
+    """
+    import asyncio
+
+    import mediapipe as mp
+
+    detector = mp.solutions.face_detection.FaceDetection(min_detection_confidence=0.5)
+    period = 1.0 / fps
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply app/src/reachy_ducky_app/embodiment/gaze.py:66]: Fixed in ca963a7. Added `if fps <= 0: raise ValueError` guard at entry, before the mediapipe import — the guard is the first observable effect, so the error message names the bad value clearly. Three new tests cover fps=0 rejection, fps=-1 rejection, and fps=0.1 accepted by the guard.

File: app/src/reachy_ducky_app/voice/openai_realtime.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        Uses ``response.create`` with ``instructions=text`` and both audio +
+        text output modalities. The SDK streams audio chunks over the
+        WebSocket; the transport layer (robot's speaker) handles playback.
+        """
+        await self._connection.response.create(
+            response={
+                "output_modalities": ["audio", "text"],
+                "instructions": text,
+            },
+        )
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply app/src/reachy_ducky_app/voice/openai_realtime.py:240]: Fixed in a0e4ff4. `speak_text` now drains connection events until `ResponseDoneEvent` (SDK's single terminal event — status disambiguates completed/cancelled/failed/incomplete) before returning. This keeps the websocket alive until the server finishes streaming audio, so the caller's `async with` only closes after the response is complete. Also handles `RealtimeErrorEvent` (logs + returns). Four new tests cover happy drain, failed-response logging, session-error logging, and interrupt-via-gather. Verified `openai` 1.109.1 has no `ResponseCancelledEvent` — brief's guess was wrong; actual SDK is simpler.

File: app/src/reachy_ducky_app/voice/openai_realtime.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        async for event in self._connection:
+            if isinstance(event, ResponseDoneEvent):
+                status = event.response.status
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply app/src/reachy_ducky_app/voice/openai_realtime.py:247]: Fixed in a68b45d + e4ab31c + 84e906d. `speak_text` now handles `ResponseAudioDeltaEvent` in the drain loop — base64-decodes the delta and awaits `SpeakerSink.play(pcm)`. The existing `ResponseDoneEvent`/`RealtimeErrorEvent` terminal handling is preserved. New pluggable `SpeakerSink` ABC with `MockSpeakerSink` for unit tests and `load_default_speaker_sink()` factory; hardware-backed `ReachySpeakerSink` tracked as follow-up issue #23.

File: app/src/reachy_ducky_app/voice/openai_realtime.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        """Iterate server events until the final user transcript arrives.
+
+        Returns the ``.transcript`` attribute of the first
+        ``conversation.item.input_audio_transcription.completed`` event.
+        Blocks until that event arrives; the SDK's async iterator handles
+        framing and keep-alives.
+        """
+        async for event in self._connection:
+            if isinstance(event, ConversationItemInputAudioTranscriptionCompletedEvent):
+                return event.transcript
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply app/src/reachy_ducky_app/voice/openai_realtime.py]: Fixed in a68b45d + e4ab31c + 84e906d. `get_user_text` now runs a mic-pump task in parallel with the event drain: for each frame from `MicSource.frames()`, it base64-encodes the PCM and calls `connection.input_audio_buffer.append(audio=b64)`. Server-VAD (the SDK default) auto-commits, so transcription fires naturally and the event arrives. The pump is cancelled in `finally` when drain exits. New pluggable `MicSource` ABC mirrors `WakeDetector` — `MockMicSource` for unit tests, `load_default_mic_source()` factory, hardware-backed `ReachyMicSource` tracked as follow-up issue #23.

**Comment by MacAttak** [comment]
@codex review + augment review — mic-in and speaker-out wired via new `MicSource` / `SpeakerSink` ABCs (commits a68b45d + e4ab31c + 84e906d). Reachy hardware-backed impls tracked in #23. E2E doc updated with the new default-mocks gap (commit 9178706). 498 tests passing, mypy strict clean.

File: app/src/reachy_ducky_app/voice/openai_realtime.py [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+            # final transcript. Let the pump run to completion so a mic
+            # failure surfaces in place of ordinary silence — without
+            # this yield the pump may never have been scheduled.
+            try:
+                await pump_task
+            except asyncio.CancelledError:
+                pass
+            return ""
+        finally:
+            if not pump_task.done():
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply app/src/reachy_ducky_app/voice/openai_realtime.py:199]: Fixed in 8146e25. `finally` block now unconditionally awaits `pump_task` even when it's already `done()`, so a stored mic exception is retrieved and re-raised instead of being orphaned. Added `_ImmediateRaisingMicSource` (raises before any `yield` / `await`, so the pump completes synchronously within its first scheduling slice) + regression test `test_get_user_text_surfaces_mic_error_over_early_transcript_return` that exercises the exact race: pump `done` with RuntimeError stored, transcription event arrives, early `return event.transcript` fires — finally's always-await surfaces the mic error over the stale transcript (correct priority for hardware failures).

**Comment by MacAttak** [comment]
@codex review + augment review — Augment medium finding (pump exception orphaned when transcription wins race) fixed in 8146e25. Full suite: 499 passed, mypy strict clean.

File: app/src/reachy_ducky_app/embodiment/gaze.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    try:
+        while True:
+            frame = mini.media.get_frame()  # type: ignore[attr-defined]
+            if frame is not None:
+                results = detector.process(frame)
+                dets: list[tuple[float, float, float]] = []
+                for d in results.detections or []:
+                    bb = d.location_data.relative_bounding_box
+                    u = bb.xmin + bb.width / 2
```
[the remark this answers is not in the document]
Comment (MacAttak) [reply app/src/reachy_ducky_app/embodiment/gaze.py]: Fixed in 716a225. Inline `min(1.0, max(0.0, ...))` clamp on both u and v before appending to the detection list, with a comment explaining the mediapipe partial-face edge case. Hardware-tier code path (no unit test added — coverage comes with #18's hardware-tier verification of ReachyMotionDriver).

**Comment by MacAttak** [comment]
@codex review + augment review — Augment low finding (mediapipe bounding-box clamping) fixed in 716a225. Codex's previous round was "Didn't find any major issues. Delightful!" — checking for any residual findings on the clamp fix.
